File: javascript/705.Design-HashSet.js
/**
 * Your MyHashSet object will be instantiated and called as such:
 * var obj = new MyHashSet()
 * obj.add(key)
 * obj.remove(key)
 * var param_3 = obj.contains(key)
 */
class Node {
  constructor(val = 0, next = null) {
    this.val = val;
    this.next = next;
  }
}

class MyHashSet {
  constructor() {
    this.max = Math.pow(10, 4);
    this.set = [];

    for (let i = 0; i < this.max; i++) {
      this.set[i] = new Node();
    }
  }

  /**
   * @param {number} key
   * @return {void}
   */
  add(key) {
    const index = key % this.max;
    let curr = this.set[index];

    while (curr && curr.next) {
      if (curr.next.val === key) {
        return;
      }

      curr = curr.next;
    }

    curr.next = new Node(key);
  }

  /**
   * @param {number} key
   * @return {void}
   */
  remove(key) {
    const index = key % this.max;
    let curr = this.set[index];

    while (curr && curr.next) {
      if (curr.next.val === key) {
        curr.next = curr.next.next;
      }

      curr = curr.next;
    }
  }

  /**
   * @param {number} key
   * @return {boolean}
   */
  contains(key) {
    const index = key % this.max;
    let curr = this.set[index];

    while (curr && curr.next) {
      if (curr.next.val === key) {
        return true;
      }

      curr = curr.next;
    }

    return false;
  }
}

File: python/705.Design-HashSet.py
class ListNode:

    def __init__(self, key = -1, val = -1, next = None):
        self.key = key
        self.val = val
        self.next = next

class MyHashSet:

    def __init__(self):
        self.size = 10000
        self.set = [ListNode() for i in range(self.size)]

    def hash(self, key: int) -> int:
        return key % self.size

    def add(self, key: int) -> None:
        hashkey = self.hash(key)
        curr = self.set[hashkey]

        while curr:
            if curr.key == key and curr.val == key:
                return
            if not curr.next:
                break

            curr = curr.next
        curr.next = ListNode(key, key)

    def remove(self, key: int) -> None:
        hashkey = self.hash(key)
        dummy = self.set[hashkey]
        prev = dummy
        curr = dummy.next

        while curr:
            if curr.key == key and curr.val == key:
                prev.next = curr.next
                return
            prev = curr 
            curr = curr.next

    def contains(self, key: int) -> bool:
        hashkey = self.hash(key)
        curr = self.set[hashkey]

        while curr:
            if curr.key == key and curr.val == key:
                return True
            curr = curr.next
        return False


# Your MyHashSet object will be instantiated and called as such:
# obj = MyHashSet()
# obj.add(key)
# obj.remove(key)
# param_3 = obj.contains(key)